--- ignition/ignition-2.16.2/internal/exec/util/unit.go
+++ ignition/ignition-2.16.2/internal/exec/util/unit.go
@@ -154,10 +154,14 @@ func (ut Util) EnableUnit(enabledUnit string) error {
 
 func (ut Util) DisableUnit(disabledUnit string) error {
 	// check if the unit is currently enabled to see if we need to disable it
-	// if it's not enabled or does not exist, we don't need to do anything
+	// if it's not enabled or does not exist, we don't need to remove the
+	// enablement symlinks.
 	args := []string{"--root", ut.DestDir, "is-enabled", disabledUnit}
 	if err := exec.Command(distro.SystemctlCmd(), args...).Run(); err != nil {
-		return nil
+		// Add the unit to a preset directive to account for cases where
+		// a unit exists, is enabled in a preset, but the presets have not
+		// been evaluated.
+		return ut.appendLineToPreset(fmt.Sprintf("disable %s", disabledUnit))
 	}
 	// We need to delete any enablement symlinks for a unit before sending it to a
 	// preset directive. This will help to disable that unit completely.
